src/network/networkd-dhcp6.c

```diff
@@ -35,6 +35,105 @@ static bool dhcp6_get_prefix_delegation(Link *link) {
                       RADV_PREFIX_DELEGATION_BOTH);
 }
 
+static bool dhcp6_has_preferred_subnet_id(Link *link) {
+        assert(link->network);
+        return link->network->router_prefix_subnet_id > -1;
+}
+
+static int dhcp6_get_preferred_delegated_prefix(
+                /* IN */
+                Manager* manager,
+                Link *link,
+                struct in6_addr *pd_prefix,
+                uint8_t pd_prefix_len,
+
+                /* OUT */
+                struct in6_addr *addr
+) {
+
+        int r;
+        union in_addr_union prefix;
+        union in_addr_union pd_prefix_union;
+        const uint8_t prefix_bits = 64 - pd_prefix_len;
+        const uint64_t n_prefixes = UINT64_C(1) << prefix_bits;
+        const int64_t subnet_id = link->network->router_prefix_subnet_id;
+        _cleanup_free_ char *assigned_buf = NULL;
+
+        pd_prefix_union.in6 = *pd_prefix;
+
+        // We start off with the original PD prefix we have been assigned and
+        // iterate from there
+        prefix.in6 = *pd_prefix;
+        log_link_debug(link, "subnet id %" PRIi64, subnet_id);
+
+        if (subnet_id > -1) { // FIXME: duplicated logic with dhcp6_has_preferred_subnet_id, probably unavoidable?
+                // If the link has a preference for a particular subnet id try to
+                // allocate that
+                if ((uint64_t)subnet_id >= n_prefixes) {
+                        log_link_debug(link, "subnet id %" PRIi64 " is out of range. Only have %" PRIu64 " subnets",
+                                        subnet_id,
+                                        n_prefixes);
+                        return -1;
```

Comment:
please return proper errors. We generally follow the style that our functions return negative errno values. Make something sensible up here, e.g. `return -ERANGE;`.

Even better, just write it like this:

```c
return log_link_debug_errno(link, SYNTHETIC_ERRNO(ERANGE), "…");
```

(the SYNTHETIC_ERRNO() macro is to indicate that we made up the error here, and didn't receive it)

Reply:
Oh! That is a good one. I though I had read the CODING_STYLE.md at some point in the past but apparently it is really worth revisiting every few years (when I contribute). Thank you for pointing this out. I really like the interface.